lib/spack/spack/binary_distribution.py

```diff
@@ -37,23 +39,307 @@
+                    # with it (but only if it has a different full_hash).
+                    if entry['mirror_url'] == mirror_url:
+                        if full_hash and full_hash != entry['spec']._full_hash:
+                            entry['spec'] = indexed_spec
+                        break
+                else:
+                    self._mirrors_for_spec[dag_hash].append({
+                        "mirror_url": mirror_url,
+                        "spec": indexed_spec,
+                    })
+        finally:
+            shutil.rmtree(tmpdir)
+
+    def get_all_built_specs(self):
+        spec_list = []
+        for dag_hash in self._mirrors_for_spec:
+            # in the absence of further information, all concrete specs
+            # with the same DAG hash are equivalent, so we can just
+            # return the first one in the list.
+            if len(self._mirrors_for_spec[dag_hash]) > 0:
+                spec_list.append(self._mirrors_for_spec[dag_hash][0]['spec'])
+
+        return spec_list
+
+    def find_built_spec(self, spec):
+        """Find the built spec corresponding to ``spec``.
+
+        If the spec can be found among the configured binary mirrors, an
+        object is returned that contains the concrete spec and the mirror url
+        where it can be found.  Otherwise, ``None`` is returned.
+
+        Args:
+            spec (Spec): Concrete spec to find
+
+        Returns:
+            An list of objects containing the found specs and mirror url where
+                each can be found, e.g.:
+
+                .. code-block:: python
+
+                    [
+                        {
+                            "spec": <concrete-spec>,
+                            "mirror_url": <mirror-root-url>
+                        }
+                    ]
+        """
+        find_hash = spec.dag_hash()
+        if find_hash not in self._mirrors_for_spec:
+            return None
+
+        return self._mirrors_for_spec[find_hash]
+
+    def update_spec(self, spec, found_list):
+        """
+        Take list of {'mirror_url': m, 'spec': s} objects and update the local
+        built_spec_cache
+        """
+        spec_dag_hash = spec.dag_hash()
+
+        if spec_dag_hash not in self._mirrors_for_spec:
+            self._mirrors_for_spec[spec_dag_hash] = found_list
+        else:
+            current_list = self._mirrors_for_spec[spec_dag_hash]
+            for new_entry in found_list:
+                for cur_entry in current_list:
+                    if new_entry['mirror_url'] == cur_entry['mirror_url']:
+                        cur_entry['spec'] = new_entry['spec']
+                        break
+                else:
+                    current_list.append = {
+                        'mirror_url': new_entry['mirror_url'],
+                        'spec': new_entry['spec'],
+                    }
+
+    def update_local_index_cache(self):
```

Comment:
I'm thinking about the use of the word "cache" here in general and how to distinguish between:

* The binary cache, i.e. the (often) remotely-hosted store of already-built Spack packages
* The index, i.e. this object

This index *does* cache files (i.e. the remote index files), so perhaps the object should be called `BinaryDistributionIndexCache` (and the docstring should mention that distinction). If you don't think that's clearer, then I'm wondering if the word "cache" should be dropped from these function names.

Reply:
Because it caches both index files and concrete specs, I had initially named it (maybe somewhat vaguely) `BinaryDistributionCacheManager`.  It does manage caches of two related kinds of things.  But the word "cache" is getting pretty heavily overloaded, so I agree it makes sense to try to clear things up to reduce potential confusion.

I put `index_cache` in that function name because it's updating the locally stored binary cache index files, as opposed to the locally stored cache of concrete specs.

I'm not vehemently opposed to either of your renaming suggestions here, but neither one feels quite right to me.  While it does cache index files, it doesn't *only* cache them, so I'm not sure I like making the class name more specific in that way.  And regarding dropping `index` from the method name `update_local_index_cache`, to me that makes the purpose of the method more ambiguous.  That method doesn't update the cache of concrete specs, only the cache of index files.

I'm worried there must be something more "overarchingly weird" in this change, making it difficult to nail down exactly what's confusing about it.  Somehow I think we need to figure out what that weirdness is.  Maybe we can sync up in person at some point to try to get to the bottom of it?